v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/cassandra/iowrapper/CassandraTableReaderFactoryCassandraIoImpl.java

```diff
@@ -0,0 +1,98 @@
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.source.reader.io.cassandra.iowrapper;
+
+import com.datastax.oss.driver.api.core.config.DriverExecutionProfile;
+import com.datastax.oss.driver.api.core.config.TypedDriverOption;
+import com.google.cloud.teleport.v2.source.reader.io.cassandra.rowmapper.CassandraSourceRowMapper;
+import com.google.cloud.teleport.v2.source.reader.io.cassandra.rowmapper.CassandraSourceRowMapperFactoryFn;
+import com.google.cloud.teleport.v2.source.reader.io.row.SourceRow;
+import com.google.cloud.teleport.v2.source.reader.io.schema.SourceSchemaReference;
+import com.google.cloud.teleport.v2.source.reader.io.schema.SourceTableSchema;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.stream.Collectors;
+import org.apache.beam.sdk.coders.SerializableCoder;
+import org.apache.beam.sdk.io.cassandra.CassandraIO;
+import org.apache.beam.sdk.io.cassandra.CassandraIO.Read;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+
+/**
+ * Generate Table Reader For Cassandra using the upstream {@link CassandraIO.Read} implementation.
+ */
+public class CassandraTableReaderFactoryCassandraIoImpl implements CassandraTableReaderFactory {
+
+  /**
+   * Returns a Table Reader for given Cassandra Source using the upstream {@link CassandraIO.Read}.
+   *
+   * @param cassandraDataSource
+   * @param sourceSchemaReference
+   * @param sourceTableSchema
+   * @return table reader for the source.
+   */
+  @Override
+  public PTransform<PBegin, PCollection<SourceRow>> getTableReader(
+      CassandraDataSource cassandraDataSource,
+      SourceSchemaReference sourceSchemaReference,
+      SourceTableSchema sourceTableSchema) {
+    CassandraSourceRowMapper cassandraSourceRowMapper =
+        getSourceRowMapper(sourceSchemaReference, sourceTableSchema);
+    DriverExecutionProfile profile =
+        cassandraDataSource.driverConfigLoader().getInitialConfig().getDefaultProfile();
+    final Read<SourceRow> tableReader =
+        CassandraIO.<SourceRow>read()
+            .withTable(sourceTableSchema.tableName())
+            .withHosts(
+                cassandraDataSource.contactPoints().stream()
+                    .map(p -> p.getHostString())
+                    .collect(Collectors.toList()))
+            .withPort(cassandraDataSource.contactPoints().get(0).getPort())
+            .withKeyspace(cassandraDataSource.loggedKeySpace())
+            .withLocalDc(cassandraDataSource.localDataCenter())
```

Comment:
How does the localDataCenter gets populated? Is is discovered  during schema discovery?

Reply:
It comes from the Config File in GCS.
Please refer to [this diff](https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/2104/files#diff-670beeda718d9ad4c92ec85debb8873ef08393852dbe2ee425023ce8aac5bb50R64-R70) from #2014